Add basic project build in CI with Github Actions

mtlynch opened the pull request:
## Changes

This adds a Github Actions step so that on every commit to the main branch or pull request, Github Actions performs a build of the project in a clean environment. This helps identify compilation or linting errors early.

I configured it to keep the `jellyfin-roku.zip` as a build artifact, as it might be a convenient way to grab a PR version for testing, but we don't strictly need it. 

As noted in the comment, the build artifact step makes a zip of a zip. We could avoid the double-zip issue by instead specifying `build/staging/` as the build artifact directory, but it feels a little brittle to duplicate the logic of creating a build zip from `.bsconfig.json`.

## Issues

None

mtlynch committed:
Add basic project build in CI with Github Actions

This adds a Github Actions step so that on every commit to the main branch or pull request, Github Actions performs a build of the project in a clean environment. This helps identify compilation or linting errors early.

cewert commented:
:heart: Very happy to see someone else who cares about CI!

I would be happy to review this for you if you're wanting to start over from scratch but may I ask why you want to do that?The [legacy workflows](https://github.com/jellyfin-archive/jellyfin-roku-legacy/tree/master/.github/workflows) were still working when they were ~~murdered~~ archived and I've been using modified versions on [my fork](https://github.com/cewert/jellyrock/tree/main/.github/workflows) that have been working for months.

We will also need to restore/recreate [separate bsconfigs](https://github.com/jellyfin-archive/jellyfin-roku-legacy/blob/master/bsconfig-prod.json) and NPM scripts for dev and prod builds. I used to use the prod CI build for publishing new releases but now I guess the maintainers just copy over the prod images manually?? You will probably want to build the prod app with CI at some point too.

mtlynch commented:
@cewert - I wasn't aware of the legacy CI workflows.

What's currently preventing them from being merged into this repo?

cewert commented:
> @cewert - I wasn't aware of the legacy CI workflows.
> 
> What's currently preventing them from being merged into this repo?

Based on my current understanding, absolutely nothing besides the current maintainers lack of interest for some unknown reason. His commit message was "Remove unused workflows, dependencies, and scripts" :shrug: 

Our app/repo used them for years successfully and users liked the automated builds. The CI linted all our code and prevented potential crashes. discovered bugs with dependencies using renovate PRs and CI. We had npm scripts to lint files locally. Bsconfig for prod builds. roku-log for logging. but all of that was [removed almost immediately](https://github.com/jellyfin/jellyfin-roku/commit/7c68c73c19a597fdc715cce450577d6490efb8b2) after @1hitsong made his fork viv. It was all removed before the v1 release. It just so happens that I wrote the majority of the stuff removed and no matter how many times I asked I couldn't get a straight answer. I didn't want to waste my time fighting them to get my PRs approved to restore all the stuff they broke so I just started working on my own fork after that.

I'd be happy to help you anyway I can - I can still approve PRs :)

mtlynch commented:
>I'd be happy to help you anyway I can - I can still approve PRs :)

I'm still kind of confused. Even if I submit a PR and you approve, isn't it still a problem if @1hitsong isn't interested in keeping CI workflows?

cewert commented:
Your safest bet would be to make an issue and discuss implementation with him if he will actually respond and talk about it. Maybe he will actually explain his reasoning to you and we can figure out a path to restoring all the things that were broken. 

If you get ignored like I do then make the PR :)

1hitsong commented:
Hi @mtlynch,

> This adds a Github Actions step so that on every commit to the main branch or pull request, Github Actions performs a build of the project in a clean environment. This helps identify compilation or linting errors early.

I've worked in enterprise software that has extensive build, linting, and testing workflows, so I understand use cases for CI; however, I'm not sure it makes sense for BrightScript because it requires so much manual work by design.

What do we gain from performing builds in GH on every commit and PR when no matter what, we must to pull the code and manually test any changes prior to merging?

mtlynch commented:
>What do we gain from performing builds in GH on every commit and PR when no matter what, we must to pull the code and manually test any changes prior to merging?

Generally, the earlier we can catch a bug, the better. 

If the IDE flags a warning as soon as you save a file, that's faster to correct than compiling, deploying to a device, and exercising the code.

As the maintainer, CI also saves you time because you don't have to bother testing PRs manually until they pass CI.

Automated tests can also catch things that are impractical to catch through manual testing. Suppose we change the behavior of [`isChainValid`](https://github.com/jellyfin/jellyfin-roku/blob/ef56494b3237f9bd8a004e7f1a5a8ded533a5e3f/source/utils/misc.bs#L358). There are 100+ calls to `isChainValid`, and we're not going to manually exercise every code path. If we have unit tests that exercise every behavior we expect from `isChainValid`, we can run the test to get confidence that changing the implementation hasn't introduced regressions.

TwitchBronBron commented:
> I'm not sure it makes sense for BrightScript because it requires so much manual work by design

You get compile and lint errors (`bsc` and `bslint`) for free since `jellyfin-roku` already depends on the BrighterScript compiler, adding plugins like bslint is simple. So I think it's hard to argue that those are in the “so much manual work by design” category.

When it comes to unit tests, I typically see the “so much manual work” argument when an app is written and designed in a way that is hard to test. The beautiful thing is there are so many small tweaks you can make to dramatically improve testability without sacrificing performance or readability. As the team gets more comfortable with testing, you naturally start writing code that’s testable from the start.

At my day job as a Roku developer, we run `bsc` and `bslint` on every PR. We have a sizable [Rooibos](https://github.com/rokucommunity/rooibos) unit test suite that runs on every PR (against [real Roku hardware](https://youtu.be/jPsL_hbJSKs?t=2835)), and a suite of [RTA](https://github.com/triwav/roku-test-automation) automation tests that run nightly. And all of this happens before our QA team does manual testing. 

I’m honestly struggling to see the validity of the argument against using CI to catch issues early. Especially for the jellyfin-roku project which has fewer resources and contributors. Having CI in place feels like a no-brainer to me. Professionally, we’ve found that it’s critical for the velocity our developers, the logical integrity of our application, and the success of our product. I've had the same findings for [all of the open source projects](https://github.com/orgs/rokucommunity/repositories?type=all) that I maintain as well.

In my opinion, these CI jobs are essential for a healthy, community-maintained open-source project because they:
- help contributors understand what’s expected and fix issues before PR review (saving maintainers’ time)
- catch mistakes made in a rush (this has saved me many times)
- help maintainers spot issues sooner and review PRs faster
- provide accountability and transparency so everyone follows the same process

I’d suggest removing the “Save build artifact” task for now, and then merging this.

_FULL DISCLOSURE: I maintain brighterscript, bslint, and rooibos. I use Jellyfin daily and want it to succeed. I think using CI is a benefit to Jellyfin-roku, pick whatever tools you want to accomplish that._

cewert commented:
> What do we gain from performing builds in GH on every commit and PR when no matter what, we must to pull the code and manually test any changes prior to merging?

The biggest advantage to automated builds is it allows more of our users to sideload and test our app. Not all users are familiar enough with programming, vscode, github, etc. to be able to setup a dev environment to deploy and test the app. The automated builds allow us to provide a link to the build when asking all our users to test the app for whatever reason. That is why we had `Sideload` instructions and links to the automated builds listed on the readme before you removed it all in v3. 4 months after the automated builds were removed, there were still [dozens of users](https://github.com/jellyfin/jellyfin-roku/issues/239#issuecomment-3152089311) trying to use and download them every week (theres no links anymore, so these were ALL bookmarks).

All of the reasons above were enough reasons for us to add automated builds years ago. The better question is, why remove them in the first place?

I noticed you started making release candidates (https://github.com/jellyfin/jellyfin-roku/releases/tag/3.0.7-rc1 https://github.com/jellyfin/jellyfin-roku/releases/tag/3.0.8-rc1 https://github.com/jellyfin/jellyfin-roku/releases/tag/3.0.9-rc1). If you had automated builds for every PR and push to main, then you could EASILY provide a link to all your users with the latest build from main - completely eliminating the need for release candidates and all the manual work involved in creating them.

cewert commented:
I forgot to even mention the fact that you are manually creating the prod release zips. This is the biggest reason you personally should want the automated builds, because then you can use the prod build zip to submit to roku for a new release and maybe more importantly, the prod release zips will be reproducible and free from human error.

1hitsong commented:
I'm still unconvinced performing a build for every single PR will be beneficial, but @mtlynch if it'll keep you contributing, then I'm in 🤘🏻 😃 

At this point, with all the subsequent comments, do you have any further updates to make to this PR?

cewert commented:
> I'm still unconvinced performing a build for every single PR will be beneficial, but @mtlynch if it'll keep you contributing, then I'm in 🤘🏻 😃

- You know the app will build successfully before wasting your time doing anything manually to test the PR.
- Users get zip files to sideload and test with [which they like](https://github.com/jellyfin/jellyfin-roku/issues/239#issuecomment-3152089311). Why is that not enough reason? 
- You know for certain that when a PR is merged it won't prevent the main branch from building if you make this a required check like it was on the legacy repo.
- You know for certain that your dependency PRs won't prevent the main branch from building.

Here's two PRs where I caught dependency bugs that prevented the app from building all because the CI checks failed. Do you not remember these or do you not consider them beneficial? 
- [brighterscript bug](https://github.com/jellyfin-archive/jellyfin-roku-legacy/pull/1396) I reported on the roku community slack channel and it was fixed the same day IIRC
- [roku-log bug](https://github.com/jellyfin-archive/jellyfin-roku-legacy/pull/1250) I fixed by [rolling back roku-log](https://github.com/jellyfin-archive/jellyfin-roku-legacy/pull/1249), which you approved.

mtlynch committed:
Remove build artifact step from Github Actions

@TwitchBronBron convinced me it's not worth the trouble: https://github.com/jellyfin/jellyfin-roku/pull/550#issuecomment-3390003984

mtlynch commented:
>I'm still unconvinced performing a build for every single PR will be beneficial, but @mtlynch if it'll keep you contributing, then I'm in 🤘🏻 😃

Full disclosure: I don't have the bandwidth to be an active contributor. I may contribute bugfixes and little improvements as time allows, but the things I like working on are things like tests, CI, and linting, which I sense you're not so excited about. 

So, I personally think this PR is beneficial, but you should merge it if you do too rather than to foster future contributions from me.

>At this point, with all the subsequent comments, do you have any further updates to make to this PR?

I took @TwitchBronBron's suggestion and removed the build artifact step, so it's now in a state where I think it's ready for merge.

1hitsong commented:
Build workflow temporarily disabled due to possible SHA1-Hulud exposure due to `@postman/tunnel-agent` in ` package-lock.json`

ref: https://helixguard.ai/blog/malicious-sha1hulud-2025-11-24

Scanner: https://github.com/standujar/sha1-hulud-scanner

cewert commented:
@1hitsong you can re-enable the workflows, we were never actually at risk in the first place :metal:

**We are not vulnerable:**
1. Our `package-lock.json` pins `@postman/tunnel-agent` to version **0.6.4** (published July 2024)
2. The compromised versions (0.6.5-0.6.7) were published November 24, 2025 and have been removed from npm
3. `npm ci` installs **exact** versions from the lockfile - it cannot install malicious versions

**The workflow itself is secure:**
- No secrets exposed to fork PRs (GitHub's default behavior)
- GITHUB_TOKEN is read-only for external contributors
- Source code is already public
